File: data-skipping-samples/MetaIndexManagerSample.py
from pyspark.sql import SparkSession
from metaindex import MetaIndexManager

if __name__ == '__main__':
	#setup SparkSession
	spark = SparkSession\
        .builder\
        .appName("MetaIndexManager Sample")\
        .getOrCreate()

	# (optional) set log level to debug specifically for metaindex package - to view the skipped objects
	log4jLogger = spark.sparkContext._jvm.org.apache.log4j
	log4jLogger.LogManager.getLogger('com.ibm.metaindex.search').setLevel(log4jLogger.Level.DEBUG)

	# setup the environment
	hconf = spark.sparkContext._jsc.hadoopConfiguration()
	# configure Stocator
	# for more info on how to config credentials see https://github.com/CODAIT/stocator
	# see https://cloud.ibm.com/docs/services/cloud-object-storage?topic=cloud-object-storage-endpoints for the list of endpoints
	# make sure you choose the private endpoint of your bucket
	hconf.set("fs.cos.service.endpoint" ,"https://s3.private.us.cloud-object-storage.appdomain.cloud")
	hconf.set("fs.cos.service.access.key", "<accessKey>")
	hconf.set("fs.cos.service.secret.key","<secretKey>")

	# dataset
	dataset_location = "cos://mybucket.service/location/to/my/data"
	# metadata base location
	md_base_location = "cos://mybucket.service/location/to/my/base/metdata"

	# Setup the JVM parameters for MetaIndexManager and enable skipping
	MetaIndexManager.injectDataSkippingRule(spark)
	# set Parquet as the default metadataStore
	MetaIndexManager.setDefaultMetaDataStore(spark, 'com.ibm.metaindex.metadata.metadatastore.parquet.Parquet')
	md_backend_config = dict([('spark.ibm.metaindex.parquet.mdlocation', md_base_location), ("spark.ibm.metaindex.parquet.mdlocation.type", "EXPLICIT_BASE_PATH_LOCATION")])
	MetaIndexManager.setConf(spark, md_backend_config)
	# Enable data skipping
	MetaIndexManager.enableFiltering(spark)

	# index the dataset
	im = MetaIndexManager(spark, dataset_location, 'com.ibm.metaindex.metadata.metadatastore.parquet.ParquetMetadataBackend')

	# remove existing index first
	if im.isIndexed():
		im.removeIndex()

	print("Building the index:")
	im.indexBuilder().addMinMaxIndex("temp").addValueListIndex("city").addBloomFilterIndex("vid").build(reader)

	# for refresh use
	#im.refreshIndex(reader)

	# to view the index status use
	im.indexStats().show()

	# Query the table and view skipping stats
	df = reader.load(dataset_location)
	df.createOrReplaceTempView("metergen")
	spark.sql("select count(*) from mtg where temp > 30").show()

	# get aggregated stats
	MetaIndexManager.getLatestQueryAggregatedStats(spark).show()

	# stop SparkSession
	spark.stop()

File: data-skipping-samples/MetaIndexManagerSample.scala
import com.ibm.metaindex.metadata.metadatastore.parquet.{Parquet, ParquetMetadataBackend}
import com.ibm.metaindex.{MetaIndexManager, Registration}
import org.apache.log4j.{Level, LogManager, Logger}
import org.apache.spark.sql.SparkSession

/**
  * Sample of indexing and querying hive table using Parquet Metadatastore
  */
object MetaIndexManagerSample {
  val logger = Logger.getLogger(this.getClass)

  def main(args: Array[String]) {
    val spark = SparkSession
      .builder()
      .appName("MetaIndexManager Sample")
      .getOrCreate()

    // (optional) set log level to debug specifically for metaindex package - to view the skipped objects
    LogManager.getLogger("com.ibm.metaindex").setLevel(Level.DEBUG)

    // Setup the environment
    // configure Stocator
    spark.sparkContext.hadoopConfiguration.set("fs.stocator.scheme.list", "cos")
    spark.sparkContext.hadoopConfiguration.set("fs.cos.impl", "com.ibm.stocator.fs.ObjectStoreFileSystem")
    spark.sparkContext.hadoopConfiguration.set("fs.stocator.cos.impl", "com.ibm.stocator.fs.cos.COSAPIClient")
    spark.sparkContext.hadoopConfiguration.set("fs.stocator.cos.scheme", "cos")

    // for more info on how to config credentials see https://github.com/CODAIT/stocator
    // see https://cloud.ibm.com/docs/services/cloud-object-storage?topic=cloud-object-storage-endpoints for the list of endpoints
    spark.sparkContext.hadoopConfiguration.set("fs.cos.service.endpoint" ,"http://s3.us-south.objectstorage.softlayer.net")
    spark.sparkContext.hadoopConfiguration.set("fs.cos.$serviceName.access.key", "<accessKey>")
    spark.sparkContext.hadoopConfiguration.set("fs.cos.service.secret.key","<secretKey>")

    // dataset location
    val input_path = "cos://mybucket.service/location/to/my/data"
    // metadata base location
    val md_base_path = "cos://mybucket.service/location/to/my/base/metdata"

    // Setup the JVM parameters for MetaIndexManager and enable skipping
    MetaIndexManager.injectDataSkippingRule(spark)
    // set Parquet as the default metadataStore
    Registration.setDefaultMetaDataStore(Parquet)
    // Set the default metadata location to be a base path defined in the database
    val jvmParameters = new java.util.HashMap[String, String]()
    jvmParameters.put("spark.ibm.metaindex.parquet.mdlocation", md_base_path)
    jvmParameters.put("spark.ibm.metaindex.parquet.mdlocation.type", "EXPLICIT_BASE_PATH_LOCATION")
    MetaIndexManager.setConf(jvmParameters)
    // Enable data skipping
    MetaIndexManager.enableFiltering(spark)

    // index the dataset
    val im = new MetaIndexManager(spark, input_path, ParquetMetadataBackend)

    // remove existing index first
    if (im.isIndexed()) {
      im.removeIndex()
    }

    // Build the index
    println("Building the index:")
    val reader = spark.read.format("parquet")
    im.indexBuilder()
      .addMinMaxIndex("temp")
      .addValueListIndex("city")
      .addBloomFilterIndex("vid")
      .build(reader)

    // for refresh use
    //im.refreshIndex(reader)

    // to view the index status use
    im.getIndexStats().show()

    // Query the table and view skipping stats
    val df = reader.load(input_path)
    df.createOrReplaceTempView("metergen")
    spark.sql("select * from metergen where temp > 30").show()

    MetaIndexManager.getLatestQueryAggregatedStats(spark).show()
  }
}
